mlflow/pyfunc/backend.py

```diff
@@ -137,33 +230,125 @@ def copy_model_into_container(dockerfile_context_dir):
         )
 
 
-def _execute_in_conda_env(conda_env_path, command, install_mlflow, command_env=None):
+def _execute_in_conda_env(conda_env_path, command, install_mlflow, env_id, command_env=None):
     if command_env is None:
         command_env = os.environ
-    env_id = os.environ.get("MLFLOW_HOME", VERSION) if install_mlflow else None
     conda_env_name = get_or_create_conda_env(conda_env_path, env_id=env_id)
     activate_conda_env = get_conda_command(conda_env_name)
     if install_mlflow:
-        if "MLFLOW_HOME" in os.environ:  # dev version
-            install_mlflow = "pip install -e {} 1>&2".format(os.environ["MLFLOW_HOME"])
-        else:
-            install_mlflow = "pip install mlflow=={} 1>&2".format(VERSION)
+        activate_conda_env += [_get_install_mlflow_command()]
+    _logger.info("=== Running command '%s'", command)
+    _run_command(
+        [
+            *_get_entrypoint(),
+            _join_commands(
+                *activate_conda_env,
+                command,
+            ),
+        ],
+        env=command_env,
+    )
 
-        activate_conda_env += [install_mlflow]
-    if os.name != "nt":
-        separator = " && "
-    else:
-        separator = " & "
 
-    command = separator.join(activate_conda_env + [command])
-    _logger.info("=== Running command '%s'", command)
+def should_use_virtualenv():
+    return "MLFLOW_USE_VIRTUALENV" in os.environ
 
-    if os.name != "nt":
-        child = subprocess.Popen(["bash", "-c", command], close_fds=True, env=command_env)
-    else:
-        child = subprocess.Popen(["cmd", "/c", command], close_fds=True, env=command_env)
-    rc = child.wait()
-    if rc != 0:
-        raise Exception(
-            "Command '{0}' returned non zero return code. Return code = {1}".format(command, rc)
+
+def _validate_pyenv_is_available():
+    # TODO
+    pass
+
+
+def _validate_virtualenv_is_available():
+    # TODO
+    pass
+
+
+def _get_env_name_for_virtualenv(requirements_file, python_env_file, env_id=None):
+    requirements_file_contents = open(requirements_file).read()
+    python_env_file_contents = open(python_env_file).read()
+    s = requirements_file_contents + python_env_file_contents + (env_id or "")
+    return "mlflow-" + hashlib.sha1(s.encode("utf-8")).hexdigest()
+
+
+def _execute_in_virtualenv(local_model_path, command, install_mlflow, env_id, command_env):
+    # Read required metadata
+    python_env_file = os.path.join(local_model_path, _PYTHON_ENV_FILE_NAME)
```

Comment:
Currently, the pyfunc Model flavors includes config like:
```
         "python_function": {
             "model_path": "model.pkl",
             "loader_module": "mlflow.sklearn",
             "python_version": "3.8.10",
             "env": "conda.yaml",
         },
```
Shall we add a key / value like "pip_requirements": "requirements.txt" ? and then we can read the requirement file name from the flavor config, instead of hardcoded `_PYTHON_ENV_FILE_NAME` .

Reply:
> we can read the requirement file name from the flavor config

What is the advantage of this approach? Can we just read it from the mode directory?

Reply:
I found "conda" env read the "env" from flavor config. shall we keep consistent with conda side  ? and also records these info in model flavor config .